sql/src/main/java/io/crate/protocols/postgres/PostgresWireProtocol.java

```diff
@@ -377,18 +383,28 @@ private void authenticate(Channel channel, Properties properties) {
             );
             Messages.sendAuthenticationError(channel, errorMessage);
         } else {
-            try {
-                User user = authMethod.authenticate(userName, connProperties);
-                if (user != null && LOGGER.isTraceEnabled()) {
-                    LOGGER.trace("Authentication succeeded user \"{}\" and method \"{}\".",
-                        user.name(), authMethod.name());
-                }
-                session = sqlOperations.createSession(properties.getProperty("database"), user);
-                Messages.sendAuthenticationOK(channel)
-                    .addListener(f -> sendParamsAndRdyForQuery(channel));
-            } catch (Exception e) {
-                Messages.sendAuthenticationError(channel, e.getMessage());
+            authContext = new AuthenticationContext(authMethod, connProperties, userName, LOGGER);
+            if ("password".equals(authMethod.name())) {
+                Messages.sendAuthenticationCleartextPassword(channel);
+                return;
             }
+            finishAuthentication(channel);
+        }
+    }
+
+    private void finishAuthentication(Channel channel) {
+        assert authContext != null : "finishAuthentication() requires an authContext instance";
+        try {
+            User user = authContext.authenticatedUser();
```

Comment:
I'd rename this method to `authenticate()`. Currently looks like a getter but it is not.

Reply:
good point